src/types.ts

```diff
@@ -51,6 +51,9 @@ export interface Battery extends OriginData {
 export interface Message {
   args: Record<string, unknown>;
```

Comment:
In the `add` case args looks like i.e.:

```js
config: "U2FsdGVkX1+jfieW172FcOgLK0KBKSLNly6zTQ...EPZjw0FRUsr1Y5HfcCyPTMMOZikwcCXm9Hsb5grS"
connector: "lnd"
name: "LND"
```

I assume in other cases this different.  
@dylancom you have an idea how to handle this?

Reply:
hmm. currently it always is `connector` (indicating which connector class to use), `name` (human readable name that we can use in the frontend), and `config` (the actual connector config; encrypted as string)

Reply:
Ah ok, that helps, will update

Reply:
- This is expecting a `.message` as well
  https://github.com/getAlby/lightning-browser-extension/blob/master/src/extension/background-script/actions/webln/signMessageOrPrompt.ts#L8
- This also has `prompt` and `origin`
  https://github.com/getAlby/lightning-browser-extension/blob/master/src/common/lib/utils.ts#L17-L22

Maybe I'm reading it wrong but I would assume this would need to be aligned?

Reply:
ah you mean the `args` - sorry somehow I was thinking about the `account` object. 

not sure what's the best typescript way for general/unknown arguments. 
in the `accounts` it needs to be come a `Account` type.

Reply:
this needs to get the correct Account type, doesn't it? https://github.com/getAlby/lightning-browser-extension/pull/636/files#diff-0029b68659877433cec643410ec71d719a7cf163141ca11e7b0ced59716530caL5

Reply:
- Yup, it's here:  
  https://github.com/getAlby/lightning-browser-extension/blob/master/src/types.ts#L51-L54  
- Could add missing props and use this I guess:  
  https://github.com/getAlby/lightning-browser-extension/blob/master/src/types.ts#L51-L54  

Just not sure if this would be correct in the overall app.

Reply:
🤔 

```diff
diff --git a/src/extension/background-script/actions/accounts/add.ts b/src/extension/background-script/actions/accounts/add.ts
index bb02736..4853212 100644
--- a/src/extension/background-script/actions/accounts/add.ts
+++ b/src/extension/background-script/actions/accounts/add.ts
@@ -1,10 +1,9 @@
 import type { Runtime } from "webextension-polyfill";
 import { encryptData } from "../../../../common/lib/crypto";
-import type { Message } from "../../../../types";
+import type { Message, Account, ConnectorType } from "../../../../types";
 import state from "../../state";

 const add = async (message: Message, _sender: Runtime.MessageSender) => {
-  const newAccount = message.args;
   const accounts = state.getState().accounts;

   // TODO: add validations
@@ -18,8 +17,11 @@ const add = async (message: Message, _sender: Runtime.MessageSender) => {
   );

   const accountId = crypto.randomUUID();
-  newAccount.config = encryptData(newAccount.config, password);
-  accounts[accountId] = newAccount;
+  accounts[accountId] = {
+    name: message.args.name as string,
+    connector: message.args.connector as ConnectorType,
+    config:  encryptData(message.args.config, password),
+  };

   state.setState({ accounts });

diff --git a/src/types.ts b/src/types.ts
index ef403a2..f97497e 100644
--- a/src/types.ts
+++ b/src/types.ts
@@ -1,7 +1,9 @@
 import connectors from "./extension/background-script/connectors";

+export type ConnectorType = keyof typeof connectors;
+
 export interface Account {
-  connector: keyof typeof connectors;
+  connector: ConnectorType;
   config: string;
   name: string;
 }
```

Reply:
I think the best way for general/unknown arguments is to use Generics: https://www.typescriptlang.org/docs/handbook/2/generics.html

We also use this for e.g. with `utils.call<Accounts>("getAccounts")` where we write what type we expect within `<>`. (libs like Axios do it also like this)

Reply:
ahhh. I see. so we define the `args` as generics here?

Reply:
Had a closer look at the code and come to the following conclusion:
First option: [narrowing-the-unknown-type](https://mariusschulz.com/blog/the-unknown-type-in-typescript#narrowing-the-unknown-type)
Second option: [using-type-assertions-with-unknown](https://mariusschulz.com/blog/the-unknown-type-in-typescript#using-type-assertions-with-unknown)

Reply:
Sorry, just noticed that my links above both pointed to the same lines... fixed it.  

> Had a closer look at the code and come to the following conclusion

I wonder if it's possible to know the exact type. The app knows what `Message` will be in each situation, no? If yes, could we define all needed types (and/or variations)?

Reply:
Yes, we could also "assume" that we pass in the correct MessageType and have an "AddAccountMessage" or something, if that's safe enough.

Reply:
THIS! 🎉   
Will come up with a suggestion for this PR.

Reply:
Thanks for tackling this @bumi .  
For further cleanup I created #652 and mentioned in related places